qiita_db/support_files/patches/python_patches/47.py

```diff
@@ -0,0 +1,677 @@
+# The code is commented with details on the changes implemented here,
+# but here is an overview of the changes needed to transfer the analysis
+# data to the plugins structure:
+# 1) Create a new type plugin to define the diversity types
+# 2) Create the new commands on the existing QIIME plugin to execute the
+#    existing analyses (beta div, taxa summaries and alpha rarefaction)
+# 3) Transfer all the data in the old structures to the plugin structures
+# 4) Delete old structures
+
+from os.path import join, exists, basename
+from os import makedirs
+from json import loads
+
+from biom import load_table, Table
+from biom.util import biom_open
+
+from qiita_db.sql_connection import TRN
+from qiita_db.util import (get_db_files_base_dir, purge_filepaths,
+                           get_mountpoint, compute_checksum)
+from qiita_db.artifact import Artifact
+
+# Create some aux functions that are going to make the code more modular
+# and easier to understand, since there is a fair amount of work to do to
+# trasnfer the data from the old structure to the new one
+
+
+def create_non_rarefied_biom_artifact(analysis, biom_data, rarefied_table):
+    """Creates the initial non-rarefied BIOM artifact of the analysis
+
+    Parameters
+    ----------
+    analysis : dict
+        Dictionary with the analysis information
+    biom_data : dict
+        Dictionary with the biom file information
+    rarefied_table : biom.Table
+        The rarefied BIOM table
+
+    Returns
+    -------
+    int
+        The id of the new artifact
+    """
+    # The non rarefied biom artifact is the initial biom table of the analysis.
+    # This table does not currently exist anywhere, so we need to actually
+    # create the BIOM file. To create this BIOM file we need: (1) the samples
+    # and artifacts they come from and (2) whether the samples where
+    # renamed or not. (1) is on the database, but we need to inferr (2) from
+    # the existing rarefied BIOM table. Fun, fun...
+
+    with TRN:
+        # Get the samples included in the BIOM table grouped by artifact id
+        # Note that the analysis contains a BIOM table per data type included
+        # in it, and the table analysis_sample does not differentiate between
+        # datatypes, so we need to check the data type in the artifact table
+        sql = """SELECT artifact_id, array_agg(sample_id)
+                 FROM qiita.analysis_sample
+                    JOIN qiita.artifact USING (artifact_id)
+                 WHERE analysis_id = %s AND data_type_id = %s
+                 GROUP BY artifact_id"""
+        TRN.add(sql, [analysis['analysis_id'], biom_data['data_type_id']])
+        samples_by_artifact = TRN.execute_fetchindex()
+
+        # Create an empty BIOM table to be the new master table
+        new_table = Table([], [], [])
+        ids_map = {}
+        for a_id, samples in samples_by_artifact:
```

Comment:
This has an extra level of complexity, right? You need to merge based on the processing params of the artifact, right?